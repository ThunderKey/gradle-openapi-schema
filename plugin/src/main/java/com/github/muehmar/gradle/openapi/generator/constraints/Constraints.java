package com.github.muehmar.gradle.openapi.generator.constraints;

import java.util.Objects;
import java.util.Optional;
import java.util.function.Consumer;

public class Constraints {
  private final Min min;
  private final Max max;
  private final DecimalMin decimalMin;
  private final DecimalMax decimalMax;
  private final Size size;
  private final Pattern pattern;
  private final Email email;

  private Constraints(
      Min min,
      Max max,
      DecimalMin decimalMin,
      DecimalMax decimalMax,
      Size size,
      Pattern pattern,
      Email email) {
    this.min = min;
    this.max = max;
    this.decimalMin = decimalMin;
    this.decimalMax = decimalMax;
    this.size = size;
    this.pattern = pattern;
    this.email = email;
  }

  public static Constraints empty() {
    return new Constraints(null, null, null, null, null, null, null);
  }

  public static Constraints ofMin(Min min) {
    return Constraints.empty().withMin(min);
  }

  public static Constraints ofMax(Max max) {
    return Constraints.empty().withMax(max);
  }

  public static Constraints ofMinAndMax(Min min, Max max) {
    return Constraints.empty().withMin(min).withMax(max);
  }

  public static Constraints ofSize(Size size) {
    return Constraints.empty().withSize(size);
  }

  public static Constraints ofPattern(Pattern pattern) {
    return Constraints.empty().withPattern(pattern);
  }

  public static Constraints ofEmail() {
    return Constraints.empty().withEmail(new Email());
  }

  public static Constraints ofDecimalMin(DecimalMin decimalMin) {
    return Constraints.empty().withDecimalMin(decimalMin);
  }

  public static Constraints ofDecimalMax(DecimalMax decimalMax) {
    return Constraints.empty().withDecimalMax(decimalMax);
  }

  public Constraints withMin(Min min) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withMax(Max max) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withSize(Size size) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withPattern(Pattern pattern) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withEmail(Email email) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withDecimalMin(DecimalMin decimalMin) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public Constraints withDecimalMax(DecimalMax decimalMax) {
    return new Constraints(min, max, decimalMin, decimalMax, size, pattern, email);
  }

  public void onMin(Consumer<Min> onMin) {
    Optional.ofNullable(min).ifPresent(onMin);
  }

  public void onMax(Consumer<Max> onMax) {
    Optional.ofNullable(max).ifPresent(onMax);
  }

  public void onDecimalMin(Consumer<DecimalMin> onDecimalMin) {
    Optional.ofNullable(decimalMin).ifPresent(onDecimalMin);
  }

  public void onDecimalMax(Consumer<DecimalMax> onDecimalMax) {
    Optional.ofNullable(decimalMax).ifPresent(onDecimalMax);
  }

  public void onSize(Consumer<Size> onSize) {
    Optional.ofNullable(size).ifPresent(onSize);
  }

  public void onPattern(Consumer<Pattern> onPattern) {
    Optional.ofNullable(pattern).ifPresent(onPattern);
  }

  public void onEmail(Consumer<Email> onEmail) {
    Optional.ofNullable(email).ifPresent(onEmail);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Constraints that = (Constraints) o;
    return Objects.equals(min, that.min)
        && Objects.equals(max, that.max)
        && Objects.equals(size, that.size)
        && Objects.equals(pattern, that.pattern)
        && Objects.equals(email, that.email);
  }

  @Override
  public int hashCode() {
    return Objects.hash(min, max, size, pattern, email);
  }

  @Override
  public String toString() {
    return "Constraints{"
        + "min="
        + min
        + ", max="
        + max
        + ", size="
        + size
        + ", pattern="
        + pattern
        + ", email="
        + email
        + '}';
  }
}
