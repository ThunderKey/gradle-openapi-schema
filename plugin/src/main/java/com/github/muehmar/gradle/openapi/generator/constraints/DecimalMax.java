package com.github.muehmar.gradle.openapi.generator.constraints;

import java.util.Objects;

public class DecimalMax {
  private final String value;
  private final boolean inclusiveMax;

  public DecimalMax(String value, boolean inclusiveMax) {
    this.value = value;
    this.inclusiveMax = inclusiveMax;
  }

  public static DecimalMax inclusive(String value) {
    return new DecimalMax(value, true);
  }

  public DecimalMax exclusiveMax() {
    return new DecimalMax(value, false);
  }

  public String getValue() {
    return value;
  }

  public boolean isInclusiveMax() {
    return inclusiveMax;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    DecimalMax that = (DecimalMax) o;
    return inclusiveMax == that.inclusiveMax && Objects.equals(value, that.value);
  }

  @Override
  public int hashCode() {
    return Objects.hash(value, inclusiveMax);
  }

  @Override
  public String toString() {
    return "DecimalMax{" + "value='" + value + '\'' + ", inclusiveMax=" + inclusiveMax + '}';
  }
}
