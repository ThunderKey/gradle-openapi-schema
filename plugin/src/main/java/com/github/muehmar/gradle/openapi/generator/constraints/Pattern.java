package com.github.muehmar.gradle.openapi.generator.constraints;

import java.util.Objects;

public class Pattern {
  private final String value;

  public Pattern(String value) {
    this.value = value;
  }

  public String getPattern() {
    return value;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Pattern pattern = (Pattern) o;
    return Objects.equals(value, pattern.value);
  }

  @Override
  public int hashCode() {
    return Objects.hash(value);
  }

  @Override
  public String toString() {
    return "Pattern{" + "value='" + value + '\'' + '}';
  }
}
