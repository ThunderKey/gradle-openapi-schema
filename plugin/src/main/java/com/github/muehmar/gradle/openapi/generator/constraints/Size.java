package com.github.muehmar.gradle.openapi.generator.constraints;

import java.util.Objects;
import java.util.Optional;

/** Size constraint of a property. Either {@code min} or {@code max} will be present (or both). */
public class Size {
  private final Integer min;
  private final Integer max;

  private Size(Integer min, Integer max) {
    this.min = min;
    this.max = max;
  }

  public static Size of(int min, int max) {
    return new Size(min, max);
  }

  public static Size ofMin(int min) {
    return new Size(min, null);
  }

  public static Size ofMax(int max) {
    return new Size(null, max);
  }

  public Optional<Integer> getMin() {
    return Optional.ofNullable(min);
  }

  public Optional<Integer> getMax() {
    return Optional.ofNullable(max);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Size size = (Size) o;
    return Objects.equals(min, size.min) && Objects.equals(max, size.max);
  }

  @Override
  public int hashCode() {
    return Objects.hash(min, max);
  }

  @Override
  public String toString() {
    return "Size{" + "min=" + min + ", max=" + max + '}';
  }
}
