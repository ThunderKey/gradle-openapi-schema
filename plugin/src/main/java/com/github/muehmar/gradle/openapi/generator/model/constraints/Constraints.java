package com.github.muehmar.gradle.openapi.generator.model.constraints;

import com.github.muehmar.gradle.openapi.util.Optionals;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Optional;
import java.util.function.Function;
import lombok.Value;

@Value
@PojoBuilder
public class Constraints {
  Optional<Min> min;
  Optional<Max> max;
  Optional<DecimalMin> decimalMin;
  Optional<DecimalMax> decimalMax;
  Optional<Size> size;
  Optional<Pattern> pattern;
  Optional<Email> email;
  Optional<PropertyCount> propertyCount;
  Optional<MultipleOf> multipleOf;

  public static Constraints empty() {
    return new Constraints(
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty(),
        Optional.empty());
  }

  public static Constraints ofMin(Min min) {
    return Constraints.empty().withMin(min);
  }

  public static Constraints ofMax(Max max) {
    return Constraints.empty().withMax(max);
  }

  public static Constraints ofMinAndMax(Min min, Max max) {
    return Constraints.empty().withMin(min).withMax(max);
  }

  public static Constraints ofSize(Size size) {
    return Constraints.empty().withSize(size);
  }

  public static Constraints ofPattern(Pattern pattern) {
    return Constraints.empty().withPattern(pattern);
  }

  public static Constraints ofEmail() {
    return Constraints.empty().withEmail(new Email());
  }

  public static Constraints ofDecimalMin(DecimalMin decimalMin) {
    return Constraints.empty().withDecimalMin(decimalMin);
  }

  public static Constraints ofDecimalMax(DecimalMax decimalMax) {
    return Constraints.empty().withDecimalMax(decimalMax);
  }

  public static Constraints ofDecimalMinAndMax(DecimalMin decimalMin, DecimalMax decimalMax) {
    return Constraints.empty().withDecimalMin(decimalMin).withDecimalMax(decimalMax);
  }

  public static Constraints ofPropertiesCount(PropertyCount propertyCount) {
    return ConstraintsBuilder.create().andOptionals().propertyCount(propertyCount).build();
  }

  public static Constraints ofMultipleOf(MultipleOf multipleOf) {
    return ConstraintsBuilder.create().andOptionals().multipleOf(multipleOf).build();
  }

  public Constraints withMin(Min min) {
    return new Constraints(
        Optional.ofNullable(min),
        max,
        decimalMin,
        decimalMax,
        size,
        pattern,
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints withMax(Max max) {
    return new Constraints(
        min,
        Optional.ofNullable(max),
        decimalMin,
        decimalMax,
        size,
        pattern,
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints withSize(Size size) {
    return new Constraints(
        min,
        max,
        decimalMin,
        decimalMax,
        Optional.ofNullable(size),
        pattern,
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints withPattern(Pattern pattern) {
    return new Constraints(
        min,
        max,
        decimalMin,
        decimalMax,
        size,
        Optional.ofNullable(pattern),
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints withEmail(Email email) {
    return new Constraints(
        min,
        max,
        decimalMin,
        decimalMax,
        size,
        pattern,
        Optional.ofNullable(email),
        propertyCount,
        multipleOf);
  }

  public Constraints withDecimalMin(DecimalMin decimalMin) {
    return new Constraints(
        min,
        max,
        Optional.ofNullable(decimalMin),
        decimalMax,
        size,
        pattern,
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints withDecimalMax(DecimalMax decimalMax) {
    return new Constraints(
        min,
        max,
        decimalMin,
        Optional.ofNullable(decimalMax),
        size,
        pattern,
        email,
        propertyCount,
        multipleOf);
  }

  public Constraints and(Constraints other) {
    return new Constraints(
        Optionals.or(min, min),
        Optionals.or(max, other.max),
        Optionals.or(decimalMin, other.decimalMin),
        Optionals.or(decimalMax, other.decimalMax),
        Optionals.or(size, other.size),
        Optionals.or(pattern, other.pattern),
        Optionals.or(email, other.email),
        Optionals.or(propertyCount, other.propertyCount),
        Optionals.or(multipleOf, other.multipleOf));
  }

  public Optional<Min> getMin() {
    return min;
  }

  public Optional<Max> getMax() {
    return max;
  }

  public Optional<DecimalMin> getDecimalMin() {
    return decimalMin;
  }

  public Optional<DecimalMax> getDecimalMax() {
    return decimalMax;
  }

  public Optional<Size> getSize() {
    return size;
  }

  public Optional<Pattern> getPattern() {
    return pattern;
  }

  public Optional<Email> getEmail() {
    return email;
  }

  public <R> Optional<R> onMinFn(Function<Min, R> onMin) {
    return min.map(onMin);
  }

  public <R> Optional<R> onMaxFn(Function<Max, R> onMax) {
    return max.map(onMax);
  }

  public <R> Optional<R> onDecimalMinFn(Function<DecimalMin, R> onDecimalMin) {
    return decimalMin.map(onDecimalMin);
  }

  public <R> Optional<R> onDecimalMaxFn(Function<DecimalMax, R> onDecimalMax) {
    return decimalMax.map(onDecimalMax);
  }

  public <R> Optional<R> onSizeFn(Function<Size, R> onSize) {
    return size.map(onSize);
  }

  public <R> Optional<R> onPatternFn(Function<Pattern, R> onPattern) {
    return pattern.map(onPattern);
  }
}
