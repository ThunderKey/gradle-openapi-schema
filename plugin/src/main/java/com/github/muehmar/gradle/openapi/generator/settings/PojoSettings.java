package com.github.muehmar.gradle.openapi.generator.settings;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojoextension.annotations.PojoExtension;
import java.io.Serializable;
import java.util.List;
import java.util.Objects;

@PojoExtension
public class PojoSettings extends PojoSettingsExtension implements Serializable {
  private final JsonSupport jsonSupport;
  private final String packageName;
  private final String suffix;
  private final boolean enableSafeBuilder;
  private final boolean enableConstraints;
  private final List<ClassTypeMapping> classTypeMappings;
  private final List<FormatTypeMapping> formatTypeMappings;
  private final EnumDescriptionSettings enumDescriptionSettings;

  PojoSettings(
      JsonSupport jsonSupport,
      String packageName,
      String suffix,
      boolean enableSafeBuilder,
      boolean enableConstraints,
      List<ClassTypeMapping> classTypeMappings,
      List<FormatTypeMapping> formatTypeMappings,
      EnumDescriptionSettings enumDescriptionSettings) {
    this.jsonSupport = jsonSupport;
    this.packageName = packageName;
    this.suffix = suffix;
    this.enableSafeBuilder = enableSafeBuilder;
    this.enableConstraints = enableConstraints;
    this.classTypeMappings = classTypeMappings;
    this.formatTypeMappings = formatTypeMappings;
    this.enumDescriptionSettings = enumDescriptionSettings;
  }

  public PojoSettings(
      JsonSupport jsonSupport,
      String packageName,
      String suffix,
      boolean enableSafeBuilder,
      boolean enableConstraints,
      PList<ClassTypeMapping> classTypeMappings,
      PList<FormatTypeMapping> formatTypeMappings,
      EnumDescriptionSettings enumDescriptionSettings) {
    this.jsonSupport = jsonSupport;
    this.packageName = packageName;
    this.suffix = suffix;
    this.enableSafeBuilder = enableSafeBuilder;
    this.enableConstraints = enableConstraints;
    this.classTypeMappings = classTypeMappings.toArrayList();
    this.formatTypeMappings = formatTypeMappings.toArrayList();
    this.enumDescriptionSettings = enumDescriptionSettings;
  }

  public String getPackageName() {
    return packageName;
  }

  public String getSuffix() {
    return suffix;
  }

  public JsonSupport getJsonSupport() {
    return jsonSupport;
  }

  public boolean isJacksonJson() {
    return jsonSupport.equals(JsonSupport.JACKSON);
  }

  public boolean isEnableConstraints() {
    return enableConstraints;
  }

  public PList<ClassTypeMapping> getClassTypeMappings() {
    return PList.fromIter(classTypeMappings);
  }

  public PList<FormatTypeMapping> getFormatTypeMappings() {
    return PList.fromIter(formatTypeMappings);
  }

  public boolean isEnableSafeBuilder() {
    return enableSafeBuilder;
  }

  public boolean isDisableSafeBuilder() {
    return !isEnableSafeBuilder();
  }

  public EnumDescriptionSettings getEnumDescriptionSettings() {
    return enumDescriptionSettings;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    PojoSettings that = (PojoSettings) o;
    return enableSafeBuilder == that.enableSafeBuilder
        && enableConstraints == that.enableConstraints
        && jsonSupport == that.jsonSupport
        && Objects.equals(packageName, that.packageName)
        && Objects.equals(suffix, that.suffix)
        && Objects.equals(classTypeMappings, that.classTypeMappings)
        && Objects.equals(formatTypeMappings, that.formatTypeMappings)
        && Objects.equals(enumDescriptionSettings, that.enumDescriptionSettings);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
        jsonSupport,
        packageName,
        suffix,
        enableSafeBuilder,
        enableConstraints,
        classTypeMappings,
        formatTypeMappings,
        enumDescriptionSettings);
  }

  @Override
  public String toString() {
    return "PojoSettings{"
        + "jsonSupport="
        + jsonSupport
        + ", packageName='"
        + packageName
        + '\''
        + ", suffix='"
        + suffix
        + '\''
        + ", enableSafeBuilder="
        + enableSafeBuilder
        + ", enableConstraints="
        + enableConstraints
        + ", classTypeMappings="
        + classTypeMappings
        + ", formatTypeMappings="
        + formatTypeMappings
        + ", enumDescriptionExtraction="
        + enumDescriptionSettings
        + '}';
  }
}
