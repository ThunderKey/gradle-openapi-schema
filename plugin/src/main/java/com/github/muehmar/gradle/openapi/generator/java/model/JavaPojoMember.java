package com.github.muehmar.gradle.openapi.generator.java.model;

import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.Name;
import com.github.muehmar.gradle.openapi.generator.model.Necessity;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import com.github.muehmar.gradle.openapi.generator.model.PojoMember;
import com.github.muehmar.gradle.openapi.generator.settings.GetterSuffixes;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class JavaPojoMember {
  private final Name name;
  private final String description;
  private final JavaType javaType;
  private final Necessity necessity;
  private final Nullability nullability;

  private JavaPojoMember(
      Name name,
      String description,
      JavaType javaType,
      Necessity necessity,
      Nullability nullability) {
    this.javaType = javaType;
    this.name = name;
    this.description = description;
    this.necessity = necessity;
    this.nullability = nullability;
  }

  public static JavaPojoMember of(
      Name name,
      String description,
      JavaType javaType,
      Necessity necessity,
      Nullability nullability) {
    return new JavaPojoMember(name, description, javaType, necessity, nullability);
  }

  public static JavaPojoMember wrap(PojoMember pojoMember, TypeMappings typeMappings) {
    final JavaType javaType = JavaType.wrap(pojoMember.getType(), typeMappings);
    return new JavaPojoMember(
        pojoMember.getName(),
        pojoMember.getDescription(),
        javaType,
        pojoMember.getNecessity(),
        pojoMember.getNullability());
  }

  public Nullability getNullability() {
    return nullability;
  }

  public Necessity getNecessity() {
    return necessity;
  }

  public String getDescription() {
    return description;
  }

  public Name getName() {
    return name;
  }

  public JavaType getJavaType() {
    return javaType;
  }

  public boolean isOptional() {
    return necessity.isOptional();
  }

  public boolean isRequired() {
    return necessity.isRequired();
  }

  public boolean isNullable() {
    return nullability.isNullable();
  }

  public boolean isNotNullable() {
    return nullability.isNotNullable();
  }

  public boolean isRequiredAndNullable() {
    return isRequired() && isNullable();
  }

  public boolean isRequiredAndNotNullable() {
    return isRequired() && isNotNullable();
  }

  public boolean isOptionalAndNullable() {
    return isOptional() && isNullable();
  }

  public boolean isOptionalAndNotNullable() {
    return isOptional() && isNotNullable();
  }

  public Name getGetterName() {
    return prefixedMethodName("get");
  }

  public Name getGetterNameWithSuffix(PojoSettings settings) {
    return getGetterName().append(determineSuffix(settings));
  }

  private String determineSuffix(PojoSettings settings) {
    final GetterSuffixes getterSuffixes = settings.getGetterSuffixes();
    if (isRequiredAndNotNullable()) {
      return getterSuffixes.getRequiredSuffix();
    } else if (isRequiredAndNullable()) {
      return getterSuffixes.getRequiredNullableSuffix();
    } else if (isOptionalAndNotNullable()) {
      return getterSuffixes.getOptionalSuffix();
    } else {
      return getterSuffixes.getOptionalNullableSuffix();
    }
  }

  public Name getWitherName() {
    return prefixedMethodName("with");
  }

  public Name prefixedMethodName(String prefix) {
    if (prefix.isEmpty()) {
      return name;
    } else {
      return name.startUpperCase().prefix(prefix);
    }
  }
}
