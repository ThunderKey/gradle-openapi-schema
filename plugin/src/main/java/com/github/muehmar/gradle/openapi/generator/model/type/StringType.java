package com.github.muehmar.gradle.openapi.generator.model.type;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.model.Type;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import java.util.function.Function;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class StringType implements Type {
  private final Format format;
  private final String formatString;
  private final Constraints constraints;

  private StringType(Format format, String formatString, Constraints constraints) {
    this.format = format;
    this.formatString = formatString;
    this.constraints = constraints;
  }

  public static StringType noFormat() {
    return ofFormat(Format.NONE);
  }

  public static StringType uuid() {
    return ofFormat(Format.UUID);
  }

  public static StringType ofFormat(Format format) {
    return new StringType(format, format.value, Constraints.empty());
  }

  public static StringType ofFormatAndValue(Format format, String formatString) {
    return new StringType(format, formatString, Constraints.empty());
  }

  public Format getFormat() {
    return format;
  }

  public String getFormatString() {
    return formatString;
  }

  public StringType withConstraints(Constraints constraints) {
    return new StringType(format, formatString, constraints);
  }

  public StringType addConstraints(Constraints constraints) {
    return new StringType(format, formatString, this.constraints.and(constraints));
  }

  @Override
  public Constraints getConstraints() {
    return constraints;
  }

  @Override
  public <T> T fold(
      Function<NumericType, T> onNumericType,
      Function<IntegerType, T> onIntegerType,
      Function<StringType, T> onStringType,
      Function<ArrayType, T> onArrayType,
      Function<BooleanType, T> onBooleanType,
      Function<ObjectType, T> onObjectType,
      Function<EnumType, T> onEnumType,
      Function<MapType, T> onMapType,
      Function<AnyType, T> onAnyType) {
    return onStringType.apply(this);
  }

  public enum Format {
    DATE("date"),
    DATE_TIME("date-time"),
    TIME("partial-time"),
    PASSWORD("password"),
    BYTE("byte"),
    BINARY("binary"),
    EMAIL("email"),
    UUID("uuid"),
    URI("uri"),
    URL("url"),
    HOSTNAME("hostname"),
    IPV4("ipv4"),
    IPV6("ipv6"),
    NONE("none"),
    OTHER("other");

    private final String value;

    Format(String value) {
      this.value = value;
    }

    public String getValue() {
      return value;
    }

    public static Format parseString(String value) {
      return PList.fromArray(values()).find(f -> f.value.equals(value)).orElse(OTHER);
    }
  }
}
