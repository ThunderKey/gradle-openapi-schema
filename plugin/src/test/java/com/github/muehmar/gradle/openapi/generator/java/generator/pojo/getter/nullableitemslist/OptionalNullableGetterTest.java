package com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter.nullableitemslist;

import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.getter.nullableitemslist.OptionalNullableGetter.optionalNullableGetter;
import static com.github.muehmar.gradle.openapi.generator.model.Necessity.OPTIONAL;
import static com.github.muehmar.gradle.openapi.generator.model.Nullability.NULLABLE;
import static com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings.defaultTestSettings;
import static com.github.muehmar.gradle.openapi.snapshot.SnapshotUtil.writerSnapshot;
import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import com.github.muehmar.gradle.openapi.generator.java.model.member.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.member.TestJavaPojoMembers;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Constraints;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Pattern;
import com.github.muehmar.gradle.openapi.generator.model.constraints.Size;
import com.github.muehmar.gradle.openapi.generator.model.type.StringType;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.api.Test;

@SnapshotTest
class OptionalNullableGetterTest {
  private Expect expect;

  @Test
  @SnapshotName("nullableStringItem")
  void generate_when_nullableStringItem_then_matchSnapshot() {
    final Generator<JavaPojoMember, PojoSettings> generator = optionalNullableGetter();
    final JavaPojoMember member =
        TestJavaPojoMembers.list(
            StringType.noFormat()
                .withNullability(NULLABLE)
                .withConstraints(Constraints.ofPattern(Pattern.ofUnescapedString("pattern"))),
            OPTIONAL,
            NULLABLE,
            Constraints.ofSize(Size.of(5, 10)));

    final Writer writer = generator.generate(member, defaultTestSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }
}
