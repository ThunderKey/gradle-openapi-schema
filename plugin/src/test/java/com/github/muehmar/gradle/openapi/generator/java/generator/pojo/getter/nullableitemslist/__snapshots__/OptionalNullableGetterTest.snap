nullableStringItem[NO_VALIDATION]=[
com.fasterxml.jackson.annotation.JsonIgnore
com.fasterxml.jackson.annotation.JsonInclude
com.fasterxml.jackson.annotation.JsonProperty
com.github.muehmar.openapi.util.JacksonNullContainer
com.github.muehmar.openapi.util.Tristate
java.util.List

@JsonIgnore
public Tristate<List<Optional<String>>> getListVal() {
  return Tristate.ofNullableAndNullFlag(wrapNullableItemsList(listVal), isListValNull);
}

@JsonProperty("listVal")
@JsonInclude(JsonInclude.Include.NON_NULL)
private Object getListValJackson() {
  return isListValNull ? new JacksonNullContainer<>(listVal) : listVal;
}

]


nullableStringItem[STANDARD]=[
com.fasterxml.jackson.annotation.JsonIgnore
com.fasterxml.jackson.annotation.JsonInclude
com.fasterxml.jackson.annotation.JsonProperty
com.github.muehmar.openapi.util.JacksonNullContainer
com.github.muehmar.openapi.util.Tristate
java.util.List
javax.validation.constraints.Pattern
javax.validation.constraints.Size

@JsonIgnore
public Tristate<List<Optional<String>>> getListVal() {
  return Tristate.ofNullableAndNullFlag(wrapNullableItemsList(listVal), isListValNull);
}

@JsonProperty("listVal")
@JsonInclude(JsonInclude.Include.NON_NULL)
private Object getListValJackson() {
  return isListValNull ? new JacksonNullContainer<>(listVal) : listVal;
}
.
.
@JsonIgnore
@Size(min = 5, max = 10)
private List<@Pattern(regexp="pattern") String> getListValRaw() {
  return listVal;
}
]